autogen/browser_utils.py

```diff
@@ -280,3 +313,741 @@ def _fetch_page(self, url: str) -> None:
+            with open(save_path, "wb") as file:
+                for chunk in response.iter_content(1024):
+                    file.write(chunk)
+
+    # Extract cookies from WebDriver
+    cookies = get_cookies(driver)
+
+    # Convert cookies for use with requests
+    session_cookies = convert_cookies_to_requests_format(cookies)
+
+    # Define the user-agent if you want to match the one used by your WebDriver
+    user_agent = driver.execute_script("return navigator.userAgent;")
+
+    # Download file using requests with the same session cookies and headers
+    download_file_with_cookies(download_url, session_cookies, save_path, user_agent=user_agent)
+
+
+def display_binary_image(binary_data):
+    """
+    display_binary_image(binary_data):
+    This function displays the binary image data in Jupyter notebook cells or shows it in non-notebook environments.
+
+    Args:
+    - binary_data (bytes): A bytes object containing the PNG image data.
+
+    Returns:
+    - Nothing, but in non-notebook environment, it displays the image.
+    """
+    img = Image.open(io.BytesIO(binary_data))
+    try:
+        __IPYTHON__
+        display(img)
+    except NameError:
+        img.show()
+
+
+def generate_png_filename(url: str):  # Function to help provide a PNG filename (with relative path)
+    """
+    Generates a PNG filename based on the provided URL, along with a small random hash.
+
+    Args:
+        url (str): The URL from which to create a filename.
+
+    Returns:
+        str: A unique PNG filename based on the URL and a random hash.
+    """
+
+    # Split the URL into its components
+    parsed_url = urlparse(url)
+
+    # Generate a 4-character random hash from lowercase letters and digits
+    random_hash = "".join(random.choices(string.ascii_lowercase + string.digits, k=6))
+
+    return f"{'.'.join(parsed_url.netloc.split('.')[-2:])}-{random_hash}.png"
+
+
+def SeleniumBrowser(**kwargs):  # Function that loads the web driver
+    """
+    This function launches a headless Selenium browser based on the specified 'browser'. The available options are 'edge', 'firefox', and 'chrome'.
+
+    Parameters:
+        browser (str): A string specifying which browser to launch. Defaults to 'firefox'.
+        download_dir (str): A path to where downloaded files are stored.  Defaults to None
+
+    Returns:
+        webdriver: An instance of the Selenium WebDriver based on the specified browser.  User can open a new page by `webdriver.get('https://www.microsoft.com')`.
+
+    Raises:
+        ImportError: If selenium package is not installed, it raises an ImportError with a message suggesting to install it using pip.
+    """
+
+    # Load the arguments from kwargs
+    browser = kwargs.get("browser", "edge")
+    download_dir = kwargs.get("download_dir", tempfile.gettempdir())
+    if not download_dir:
+        download_dir = tempfile.gettempdir()
+
+    browser_res = kwargs.get("resolution", (1920, 5200))
```

Comment:
Recent updates for screenshots either resize the window to fit the rendered content or, in the case of Firefox, use the full page screenshot.  It may be more memory efficient for all users to have a standard 1080 height.